File: sdk/python/examples/33_single_turn_tool.py
# Copyright (c) 2025 Agentspan
# Licensed under the MIT License. See LICENSE file in the project root for details.

"""Single-Turn Tool Call — LLM calls a tool and answers in one shot.

The simplest tool-calling pattern: the user asks a question, the LLM
calls a tool to get data, then responds with the answer.  No iterative
loop — the agent runs for exactly one exchange.

Compiled workflow:

    LLM(prompt, tools) → tool executes → LLM sees result → answer

Requirements:
    - Conductor server with LLM support
    - AGENTSPAN_SERVER_URL=http://localhost:8080/api in .env or environment
    - AGENT_LLM_MODEL=openai/gpt-4o-mini in .env or environment
"""

from agentspan.agents import Agent, AgentRuntime, tool
from settings import settings


@tool
def get_weather(city: str) -> dict:
    """Get the current weather for a city."""
    return {"city": city, "temp_f": 72, "condition": "Sunny"}


agent = Agent(
    name="weather_agent",
    model=settings.llm_model,
    instructions="You are a weather assistant. Use the get_weather tool to answer.",
    tools=[get_weather],
    max_turns=2,  # 1 turn to call the tool, 1 turn to answer
)

with AgentRuntime() as runtime:
    result = runtime.run(agent, "What's the weather in San Francisco?")
    result.print_result()

File: sdk/typescript/examples/33-single-turn-tool.ts
/**
 * 33 - Single-Turn Tool Call
 *
 * The simplest tool-calling pattern: the user asks a question, the LLM
 * calls a tool to get data, then responds with the answer. No iterative
 * loop -- the agent runs for exactly one exchange.
 *
 * Compiled workflow:
 *   LLM(prompt, tools) -> tool executes -> LLM sees result -> answer
 *
 * Requirements:
 *   - Conductor server with LLM support
 *   - AGENTSPAN_SERVER_URL=http://localhost:8080/api as environment variable
 *   - AGENTSPAN_LLM_MODEL=openai/gpt-4o-mini as environment variable
 */

import { z } from 'zod';
import { Agent, AgentRuntime, tool } from '../src/index.js';
import { llmModel } from './settings.js';

const getWeather = tool(
  async (args: { city: string }) => {
    return { city: args.city, temp_f: 72, condition: 'Sunny' };
  },
  {
    name: 'get_weather',
    description: 'Get the current weather for a city.',
    inputSchema: z.object({
      city: z.string().describe('City name'),
    }),
  },
);

export const agent = new Agent({
  name: 'weather_agent',
  model: llmModel,
  instructions: 'You are a weather assistant. Use the get_weather tool to answer.',
  tools: [getWeather],
  maxTurns: 2, // 1 turn to call the tool, 1 turn to answer
});

// Only run when executed directly (not when imported for discovery)
async function main() {
  const runtime = new AgentRuntime();
  try {
    // Deploy to server. CLI alternative (recommended for CI/CD):
    //   agentspan deploy <module>
    await runtime.deploy(agent);
    await runtime.serve(agent);

    // Quick test: uncomment below (and comment out serve) to run directly.
    // const runtime = new AgentRuntime();
    // try {
    // const result = await runtime.run(agent, "What's the weather in San Francisco?");
    // result.printResult();
  } finally {
    await runtime.shutdown();
    // }
}

if (process.argv[1]?.endsWith('33-single-turn-tool.ts') || process.argv[1]?.endsWith('33-single-turn-tool.js')) {
  main().catch(console.error);
}
